src/server_name_resolver.rs

```diff
@@ -0,0 +1,44 @@
+use std::error::Error;
+
+use http::Uri;
+use pki_types::ServerName;
+
+/// A trait implemented by types that can resolve a [`ServerName`] for a request.
+pub trait ResolveServerName {
+    /// Maps a [`Uri`] into a [`ServerName`].
+    fn resolve(&self, uri: &Uri) -> Result<ServerName<'static>, Box<dyn Error + Sync + Send>>;
+}
+
+impl<F, E> ResolveServerName for F
+where
+    F: Fn(&Uri) -> Result<ServerName<'static>, E>,
+    E: Into<Box<dyn Error + Sync + Send>>,
+{
+    fn resolve(&self, uri: &Uri) -> Result<ServerName<'static>, Box<dyn Error + Sync + Send>> {
+        self(uri).map_err(Into::into)
+    }
+}
+
+pub(crate) struct FixedServerNameResolver {
+    name: String,
```

Comment:
Ideally we'd hold a `ServerName` in here so that we don't have to do the conversion on every connection attempt, but I guess that's tricky to fit in a semver-compatible API. I suppose we can fix that with the next semver-compatible version.

Maybe we should just `#[deprecate]` `with_server_name()` (and make this `pub`)?

Reply:
Yep, I'll update this to take in a `ServerName` directly, and then deprecate with_server_name and update it to just use a 1-off closure.